src/table/iterator.rs

```diff
@@ -0,0 +1,179 @@
+
+#[derive(Clone, Debug)]
+pub enum IteratorError {
+    NoError,
+    EOF,
+    // TODO: As we need to clone Error from block iterator to table iterator,
+    // we had to save `crate::Error` as String. In the future, we could let all
+    // seek-related function to return a `Result<()>` instead of saving an
+    // error inside struct.
+    Error(String),
+}
+
+impl IteratorError {
+    pub fn is_err(&self) -> bool {
+        !matches!(self, IteratorError::NoError)
+    }
+
+    pub fn is_eof(&self) -> bool {
+        matches!(self, IteratorError::EOF)
+    }
+}
+
+enum SeekPos {
+    Origin,
+    Current,
+}
+
+/// Block iterator iterates on an SST block
+// TODO: support custom comparator
+struct BlockIterator {
+    idx: isize,
+    base_key: Bytes,
+    key: BytesMut,
+    val: Bytes,
+    data: Bytes,
+    // TODO: use `&'a Block` if possible
+    block: Arc<Block>,
+    perv_overlap: u16,
+    err: IteratorError,
+}
+
+impl BlockIterator {
+    pub fn new(block: Arc<Block>) -> Self {
+        let data = block.data.slice(..block.entries_index_start);
+        Self {
+            block,
+            err: IteratorError::NoError,
+            base_key: Bytes::new(),
+            key: BytesMut::new(),
+            val: Bytes::new(),
+            data,
+            perv_overlap: 0,
+            idx: 0,
+        }
+    }
+
+    #[inline]
+    fn entry_offsets(&self) -> &[u32] {
+        &self.block.entry_offsets
+    }
+    fn set_idx(&mut self, i: isize) {
+        self.idx = i;
+        if i >= self.entry_offsets().len() as isize || i < 0 {
+            self.err = IteratorError::EOF;
+            return;
+        }
+
+        self.err = IteratorError::NoError;
+        let start_offset = self.entry_offsets()[i as usize] as u32;
+
+        if self.base_key.is_empty() {
+            let mut base_header = Header::default();
+            base_header.decode(&mut self.data.slice(..));
+            // TODO: combine this decode with header decode to avoid slice ptr copy
+            self.base_key = self
+                .data
+                .slice(HEADER_SIZE..HEADER_SIZE + base_header.diff as usize);
+        }
+
+        let end_offset = if self.idx + 1 == self.entry_offsets().len() as isize {
+            self.data.len()
+        } else {
+            self.entry_offsets()[self.idx as usize + 1] as usize
+        };
+
+        let mut entry_data = self.data.slice(start_offset as usize..end_offset as usize);
+        let mut header = Header::default();
+        header.decode(&mut entry_data);
+
+        // TODO: merge this truncate with the following key truncate
+        if header.overlap > self.perv_overlap {
+            self.key.truncate(self.perv_overlap as usize);
+            self.key.extend_from_slice(
+                &self.base_key[self.perv_overlap as usize..header.overlap as usize],
+            );
+        }
+        self.perv_overlap = header.overlap;
+
+        let diff_key = &entry_data[..header.diff as usize];
```

Comment:
where is `valueOff`?

Reply:
As `header.decode(&mut entry_data);` eats headers from `entry_data`, at this point, `entry_data` won't contain header. So we don't need to introduce `valueOff`.